packages/shorebird_cli/test/src/commands/preview_command_test.dart

```diff
@@ -1267,6 +1278,376 @@ channel: ${DeploymentTrack.staging.channel}
+
+      const androidReleaseArtifactUrl = 'https://example.com/release.aab';
+      const androidPackageName = 'com.example.app';
+
+      late ReleaseArtifact iosReleaseArtifact;
+      late ReleaseArtifact androidReleaseArtifact;
+
+      late Adb adb;
+      late Bundletool bundletool;
+      late Process process;
+
+      R runWithOverrides<R>(R Function() body) {
+        return HttpOverrides.runZoned(
+          () => runScoped(
+            body,
+            values: {
+              adbRef.overrideWith(() => adb),
+              artifactManagerRef.overrideWith(() => artifactManager),
+              authRef.overrideWith(() => auth),
+              bundletoolRef.overrideWith(() => bundletool),
+              cacheRef.overrideWith(() => cache),
+              codePushClientWrapperRef
+                  .overrideWith(() => codePushClientWrapper),
+              devicectlRef.overrideWith(() => devicectl),
+              httpClientRef.overrideWith(() => httpClient),
+              iosDeployRef.overrideWith(() => iosDeploy),
+              loggerRef.overrideWith(() => logger),
+              platformRef.overrideWith(() => platform),
+              shorebirdEnvRef.overrideWith(() => shorebirdEnv),
+              shorebirdValidatorRef.overrideWith(() => shorebirdValidator),
+            },
+          ),
+        );
+      }
+
+      setUp(() {
+        adb = MockAdb();
+        bundletool = MockBundleTool();
+        process = MockProcess();
+        devicectl = MockDevicectl();
+        iosDeploy = MockIOSDeploy();
+
+        when(() => appleDevice.name).thenReturn('iPhone 12');
+        when(() => appleDevice.udid).thenReturn('12345678-1234567890ABCDEF');
+        when(
+          () => artifactManager.downloadFile(any()),
+        ).thenAnswer((_) async => File(''));
+        when(
+          () => artifactManager.extractZip(
+            zipFile: any(named: 'zipFile'),
+            outputDirectory: any(named: 'outputDirectory'),
+          ),
+        ).thenAnswer((_) async {});
+
+        when(() => devicectl.deviceForLaunch(deviceId: any(named: 'deviceId')))
+            .thenAnswer((_) async => null);
+        when(
+          () => devicectl.installAndLaunchApp(
+            runnerAppDirectory: any(named: 'runnerAppDirectory'),
+            device: any(named: 'device'),
+          ),
+        ).thenAnswer((_) async => ExitCode.success.code);
+        when(() => iosDeploy.installIfNeeded()).thenAnswer((_) async {});
+        when(
+          () => iosDeploy.installAndLaunchApp(
+            bundlePath: any(named: 'bundlePath'),
+            deviceId: any(named: 'deviceId'),
+          ),
+        ).thenAnswer((_) async => ExitCode.success.code);
+        when(() => release.platformStatuses).thenReturn({
+          ReleasePlatform.android: ReleaseStatus.active,
+          ReleasePlatform.ios: ReleaseStatus.active,
+        });
+
+        iosReleaseArtifact = MockReleaseArtifact();
+        androidReleaseArtifact = MockReleaseArtifact();
```

Comment:
Prefer to instantiate mocks in a single block (these should be moved up with the other mock constructors)